### PR TITLE
Unroll do in Scribble protocol

### PR DESCRIPTION
**Pull request by fangyi-zhou**

### CODE REVIEW

File: lib/gtype.ml [hunk 1]
```diff
@@ -79,21 +77,10 @@ let of_protocol global_protocol =
...
             ( role
             , List.map ~f:(conv_interactions rec_names) interactions_list )
-      | Do (name_, _, roles_, _)
-        when String.equal name name_ && List.equal String.equal roles roles_
-        ->
-          has_global_recursion := true ;
-          assert_empty rest ;
-          TVarG global_recursion_name
-      | Do _ -> unimpl "handle `do` that are non-identical to top level"
+      | Do _ -> assert false
```
[the remark this answers is not in the document]
Comment (fangyi-zhou) [reply lib/gtype.ml]: Yes

[the remark this answers is not in the document]
Comment (fangyi-zhou) [reply lib/gtype.ml]: Actually, do we want a GADT or sth to represent different stages of AST?

[the remark this answers is not in the document]
Comment (fferreira) [reply lib/gtype.ml]: Oh my, do we? I want to say yes. But at the same time I am afraid of it becoming painful down the road.

[the remark this answers is not in the document]
Comment (fferreira) [reply lib/gtype.ml]: I will accept this, we can always GADT it later

[the remark this answers is not in the document]
Comment (fferreira) [reply lib/gtype.ml]: Well, after you add the violation

[the remark this answers is not in the document]
Comment (fferreira) [reply lib/gtype.ml]: I did it

**Pull request merged**